pandas/core/internals/blocks.py

```diff
@@ -424,7 +426,21 @@ def coerce_to_target_dtype(self, other) -> Block:
         and will receive the same block
         """
         new_dtype = find_result_type(self.values, other)
-
+        if warn_on_upcast:
+            warnings.warn(
+                f"Setting an item of incompatible dtype is deprecated "
+                "and will raise in a future error of pandas. "
+                f"Value '{other}' has dtype incompatible with {self.dtype}, "
+                "please explicitly cast to a compatible dtype first.",
```

Comment:
Thinking aloud. It would be difficult to suggest which dtype to cast? I guess `object` would be safest though

Reply:
in practice i think it will almost always be x->object or int->float

Reply:
maybe - but I'd suggest doing this separately if it's OK, this has already been open since January 😄 reckon we can get this in for 2.1? It should be pretty quick to rebase #53405, can do that right after this one

Reply:
Yep this would be fine as a follow up

Reply:
Works for me